src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb

```diff
@@ -320,31 +320,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
                 Return Nothing
             End Function
 
-            Public Overrides Function VisitAssembly(symbol As AssemblySymbol) As Symbol
-                If symbol.IsLinked Then
-                    Return symbol
+            Public Overrides Function VisitAssembly(assembly As AssemblySymbol) As Symbol
+                If assembly.IsLinked Then
+                    Return assembly
                 End If
 
-                ' the current source assembly:
-                If symbol.Identity.Equals(_sourceAssembly.Identity) Then
+                ' When we map synthesized symbols from previous generations to the latest compilation 
+                ' we might encounter a symbol that is defined in arbitrary preceding generation, 
+                ' not just the immediately preceding generation. If the source assembly uses time-based 
+                ' versioning assemblies of preceding generations might differ in their version number.
+                If IdentityEqualIgnoringVersionWildcard(assembly, _sourceAssembly) Then
                     Return _otherAssembly
                 End If
 
                 ' find a referenced assembly with the exactly same source identity:
                 For Each otherReferencedAssembly In _otherAssembly.Modules(0).ReferencedAssemblySymbols
-                    Dim identity = symbol.Identity
-                    Dim otherIdentity = otherReferencedAssembly.Identity
-
-                    If AssemblyIdentityComparer.SimpleNameComparer.Equals(identity.Name, otherIdentity.Name) AndAlso
-                       If(symbol.AssemblyVersionPattern, symbol.Identity.Version).Equals(If(otherReferencedAssembly.AssemblyVersionPattern, symbol.Identity.Version)) AndAlso
-                       AssemblyIdentity.EqualIgnoringNameAndVersion(identity, otherIdentity) Then
+                    If IdentityEqualIgnoringVersionWildcard(assembly, otherReferencedAssembly) Then
                         Return otherReferencedAssembly
                     End If
                 Next
 
                 Return Nothing
             End Function
 
+            Private Shared Function IdentityEqualIgnoringVersionWildcard(left As AssemblySymbol, right As AssemblySymbol) As Boolean
+                Dim leftIdentity = left.Identity
+                Dim rightIdentity = right.Identity
+                Return AssemblyIdentityComparer.SimpleNameComparer.Equals(leftIdentity.Name, rightIdentity.Name) AndAlso
+                       If(left.AssemblyVersionPattern, leftIdentity.Version).Equals(If(right.AssemblyVersionPattern, rightIdentity.Version)) AndAlso
+                       AssemblyIdentity.EqualIgnoringNameAndVersion(leftIdentity, rightIdentity)
+            End Function
```

Comment:
We need AssemblyVersionPattern property from the symbols. We could make this shared but it's not trivial so I'd rather keep it as is for U3.